Microverse. Project: Testing Your Ruby Code - Tic Tac Toe

AlejandroNo4 opened the pull request:
This project is to check, using RSpec, the correct functionality of projects previously built with Ruby.
The instructions said that there must be tests for every method created on the tic tac toe project.

There is just one method that doesn't have a test. It is because the ***method*** is just an infinite loop, and it's going to be calling the ***wrong_message*** and the ***character validator method*** until the user inputs a valid character.
I tried to find answers on how to make this test using the best practices. I didn't find a clear answer. It almost looks like it is subjective. What I did, was decompose and rearrange that method. Then, I created another one. In this way, I can test the ***content of the method*** without running the loop. 

Another detail is that temporarily, the code is not running the game with ***main.rb***. I cut off the ***calling line*** because when RSpec it's called, the game starts. I'm not sure if that's necessary because it is not the main objective of the project.

Thank you for your time.

AlejandroNo4 committed:
Added first files to rspec

AlejandroNo4 committed:
Added the rspec tests

AlejandroNo4 committed:
Fixed some rubocop ofences

AlejandroNo4 committed:
fixed typo and rubocop errors

rindrajosia commented on spec/tictactoe_spec.rb:
```diff
@@ -0,0 +1,118 @@
+# spec/test_tictactoe.rb
+require '../lib/core_game'
+require '../bin/main'
```
If you remove this and add `game_setting` in your `main.rb` the game will not run when you test your code.

rindrajosia commented on spec/tictactoe_spec.rb:
```diff
@@ -0,0 +1,118 @@
+      it 'puts tie_message' do
+        test_game.i = 9
+        expect(test_game.switch).to eql(tie_message)
+      end
```
This test is not necessary as it will just display a message. Instead, you **need just to test if there is a** `winner` or `tie`.

rindrajosia reviewed (changes requested):
**CHANGES REQUIRED**
Hello, you've done a great job working on this project. However, there is a very small issue you need to do to make the game work properly.

- [x] You don't need to test the method inside the `main.rb` file as some of them are there to display a message, Instead, you need just to test if you have a winning condition or a tie condition.
- [x] You need also test negative scenarios, you might use "to_not eql" and "to_not be_an".
- [x] The game should work even in this milestone.

Good luck and happy coding
Rindra Josia

AlejandroNo4 committed:
Added the required changes

AlejandroNo4 commented:
Here is the second version.
Here I added the required changes. The principal is that the "tie" and "winning" conditionals are at the final of the code.

The full game is a loop of nine rounds, and if there isn't a winner at the 9nth, the game ends with a tie. The only way to "break" this loop is when the variable ***test_res*** has a truth value. In this way, if it's added the final possible "spot," and the ***test_res*** stays as false, the loop ends, and the result will be a tie.

I also followed the recommendation, cutting off the switch method. This method is the loop that calls the messages.

I added the negative scenarios, but in parts where there are positive ones, but not the same, so the code is not repetitive.

mke2111 reviewed (approved):
# **STATUS: APPROVED**

**Congratulations Alejandro, The project meets most of the requirements. Thanks for the splendid job.**

- [x] [Optional] Using `require_relative` for files in your directory instead of `require` is a ruby convention. The latter is used for external files like gems. Always remember that and make changes to your files.

**You can now merge to the develop branch and submit a project completion form.
Best wishes in your upcoming projects, and happy coding.**

### Cheers!!
![image](https://user-images.githubusercontent.com/65501535/105930052-f1cc8380-6059-11eb-8b03-89951ccdbce5.png)

**Reach me on slack.
(Roy Mukuye)**
